### DISCUSSION TITLE
Get rid of all VC++ 2013 Redistributable

### DISCUSSION THREAD

**Discussion by VaclavElias** [stride3d/stride#1498] [Ideas]
# Motivation: Making Build and Install process/steps bulletproof

I have put this together, if somebody would be doing the same research or would like to take over this..

I found this [comment](https://discord.com/channels/500285081265635328/500290856532574209/630858318008811541) from @xen2 from 2019.

![image](https://user-images.githubusercontent.com/4528464/183302106-105c6f33-b524-4121-96b3-3b46f2d3ad3a.png)

A few more comments from @manio143 in Discord (and here https://github.com/stride3d/stride/issues/856), that we need to install **Microsoft Visual C++ 2013 Redistributable** both x86 and x64. We do need them currently but we don't even mention it in the https://github.com/stride3d/stride Build docs, most likely because we assume we have got it already if we do some serious dev/game development.

I am doing lots of testing for code only approach + clean Windows 10 installation and would like to simplify the installation process.

Currently, if you don't install **Microsoft Visual C++ 2013 Redistributable** you can `dotnet build` and run project through command line. But when you use Visual Studio, then you can't build the project because it starts complaining because of this:

Firstly checking for **_StrideCheckVisualCRuntime2013** and then if both redistributable are installed.

```c:\Users\Vacla\.nuget\packages\stride.core.assets.compilerapp\4.1.0.1734\buildTransitive\Stride.Core.Assets.CompilerApp.targets```

```
<Project InitialTargets="_StrideCheckVisualCRuntime2013" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
  <!-- Check if Visual C++ Runtime 2013 is properly installed -->
  <!-- Should be moved to Stride.Assets as soon as transitive build dependencies are enabled by default in VS2019 (https://github.com/NuGet/Home/issues/6091#issuecomment-438073285) -->
  <Target Name="_StrideCheckVisualCRuntime2013" Condition="'$(MSBuildRuntimeType)' == 'Full'">
    <ItemGroup>
      <_StrideVisualCRuntime2013 Include="Visual C++ Redistributable for Visual Studio 2013 x86">
        <Version>$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\DevDiv\VC\Servicing\12.0\RuntimeMinimum', 'Version', null, RegistryView.Registry32))</Version>
        <ExpectedVersion>12.0.21005</ExpectedVersion>
        <DownloadUrl>http://download.microsoft.com/download/2/E/6/2E61CFA4-993B-4DD4-91DA-3737CD5CD6E3/vcredist_x86.exe</DownloadUrl>
      </_StrideVisualCRuntime2013>
      <_StrideVisualCRuntime2013 Include="Visual C++ Redistributable for Visual Studio 2013 x64">
        <Version>$([MSBuild]::GetRegistryValueFromView('HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\DevDiv\VC\Servicing\12.0\RuntimeMinimum', 'Version', null, RegistryView.Registry64))</Version>
        <ExpectedVersion>12.0.21005</ExpectedVersion>
        <DownloadUrl>http://download.microsoft.com/download/2/E/6/2E61CFA4-993B-4DD4-91DA-3737CD5CD6E3/vcredist_x64.exe</DownloadUrl>
      </_StrideVisualCRuntime2013>

      <_StrideVisualCRuntime2013NotInstalled Include="@(_StrideVisualCRuntime2013)" Condition="'%(_StrideVisualCRuntime2013.Version)' == '' Or $([System.Version]::Parse('%(Version)').CompareTo($([System.Version]::Parse('%(_StrideVisualCRuntime2013.ExpectedVersion)')))) &lt; 0" />
    </ItemGroup>

    <Error Condition="'@(_StrideVisualCRuntime2013NotInstalled)' != ''" Text="@(_StrideVisualCRuntime2013NotInstalled->'%(Identity) is not installed. Please download from %(DownloadUrl)', '%0a')"/>
  </Target>
...
</Project> 
```
Example for error in VS for code only example or FirstPersonShooter template example. I can build through the command line but not through VS.

![image](https://user-images.githubusercontent.com/4528464/183305496-22650528-fbac-4064-babb-02ad68fe4ff0.png)

I am getting lots of hints from @ericwj and I removed the ```<Target></Target>``` and I was able to build & run project through command line only with the latest package (Microsoft Visual C++ 2015-2022 Redistributable [vcredist_x64.exe](https://aka.ms/vs/17/release/vc_redist.x64.exe) (25MB))

Including Visual Studio doesn't complain anymore about 2013 version.

I tried also our template FirstPersonShooter and all was working.

Clean Win 10 Installed Software
![image](https://user-images.githubusercontent.com/4528464/183302803-e08f3484-7fc8-46e5-8e77-3fa29373f474.png)

My code only approach currently doesn't touch all Stride NuGets, so I don't know the dependencies on 2013 Redistributable. 

My question is: Do we know exactly which native library is dependant on VC++ 2013 Redistributable?

I can see many **DllImport** references in the Stride solution but not sure how to find easily the one dependant on the above.

My naive assumption is that if I was able to `dotnet build`, `dotnet run` FirstPersonShooter without VC++ 2013 Redistributable installed, do we actually need it? Obviously I don't know about the deep dependencies.

I also tried to run **Stride.GameStudio**, and it didn't complain when debugging FirstPersonShooter template loading and building.

Thanks for any comments on this.

**Comment by xen2** [comment]
Sorry I don't have a recent checkout on Stride available right now, but to get a quick idea of dependencies:
- Download DependencyWalker (x64)
- Build GameStudio
- Go in the GameStudio build dir, and open all the .dll from the x64 subdirectory to check their dependencies.

I remember VAHCD was one of the only dll still referencing MSVC2013 (there were many more but I think many were updated).
Since it's only VAHCD, I am not surprised everything might work to a certain point, as long as you don't build a sample/project requiring it.

Note: even if not used directly as an asset, it might be loaded at some point by GameStudio and/or asset compiler if Physics is used (not sure about that).

**Comment by VaclavElias** [comment]
I am using https://github.com/lucasg/Dependencies which is a rewrite of DependecyWalker...

I shall check this folder d:\Projects\GitHub\stride\sources\editor\Stride.Editor\bin\Debug\net6.0-windows7.0\?
- win-x64
- x64

And look for something which belongs to 2013 Redistributable. Shall I be looking for files *****120.dll** (https://gist.github.com/RDCH106/40fe61f447df58c1b9c83a1781374bcd)

![image](https://user-images.githubusercontent.com/4528464/183311900-38a1c5ae-4475-4159-bbe4-241862270329.png)
